common/authorization/interceptor.go

```diff
@@ -45,42 +50,79 @@ func (a *interceptor) Interceptor(
 	handler grpc.UnaryHandler,
 ) (interface{}, error) {
 
-	var namespace string
-	requestWithNamespace, ok := req.(requestWithNamespace)
-	if ok {
-		namespace = requestWithNamespace.GetNamespace()
-	}
+	var caller *Claims
 
-	apiName := info.FullMethod
-	index := strings.LastIndex(apiName, "/")
-	if index > -1 {
-		apiName = apiName[index+1:]
+	if a.claimMapper != nil {
+		var tlsSubject *pkix.Name
+		var authHeaders []string
+
+		if md, ok := metadata.FromIncomingContext(ctx); ok {
+			authHeaders = md["authorization"]
+		}
+		if p, ok := peer.FromContext(ctx); ok {
+			if tlsAuth, ok := p.AuthInfo.(credentials.TLSInfo); ok {
+				if len(tlsAuth.State.VerifiedChains) > 0 && len(tlsAuth.State.VerifiedChains[0]) > 0 {
+					tlsSubject = &tlsAuth.State.VerifiedChains[0][0].Subject
+				}
+			}
+		}
+		// Add auth into to ctx only if there's some auth info
+		if tlsSubject != nil || len(authHeaders) > 0 {
+			authInfo := AuthInfo{authHeaders[0], tlsSubject}
+			if a.authorizer != nil {
+				claims, err := a.claimMapper.GetClaims(authInfo)
+				if err != nil {
+					return nil, err
+				}
+				caller = claims
+				ctx = context.WithValue(ctx, "auth-claims", claims)
+			}
+		}
 	}
 
-	scope := a.getMetricsScope(metrics.NumAuthorizationScopes, namespace)
-	sw := scope.StartTimer(metrics.ServiceAuthorizationLatency)
-	defer sw.Stop()
+	if a.authorizer != nil {
+		var namespace string
+		requestWithNamespace, ok := req.(requestWithNamespace)
+		if ok {
```

Comment:
what do we do if namespace is empty? Does that impact the initialization of the metrics scope down below? Should we have a default value then if ok is false?

Reply:
It is handled explicitly by the metrics scope initialization code - https://github.com/sergeybykov/temporal/blob/830bdd53498bf377afa638d79ade8e6f5c844d5f/common/authorization/interceptor.go#L137.
I think it's okay to use "" namespace for operations that are at the cluster level. Unless there's a better way to do that.

Reply:
just thinking about query experience - as long as there is a way to explicitly search for cluster-level operations using metrics (in this case, it looks like namespace=""), then I think thats fine. Seems it might be easier if we have a special tag value though.

Reply:
Do you mean a special tag different from `metrics.NamespaceUnknownTag`?

Reply:
> what do we do if namespace is empty?

should we treat empty namespace just like other normal namespace to reduce number of edge case?

Reply:
I think this question goes beyond the scope of this PR. For authZ though, we can't treat "" as just another namespace because giving access permissions to it the same way as to other namespaces doesn't make sense.

Reply:
Discuss with @wxing1292 offline. This logic will evolve later in subsequent changes.